File: Coursera/Algorithms_Part_2/Ex16PrimsLazyMST.java
/* *****************************************************************************
 *  Name:
 *  Date:
 *  Description:
 **************************************************************************** */

import edu.princeton.cs.algs4.Bag;
import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.MinPQ;
import edu.princeton.cs.algs4.StdOut;

public class Ex16PrimsLazyMST implements Ex14MST {

    private Bag<Ex12Edge> edges;
    private MinPQ<Ex12Edge> pq;
    private boolean[] marked;
    private double weight;

    public Ex16PrimsLazyMST(Ex13EdgeWeightedGraph G) {
        edges = new Bag<>();
        pq = new MinPQ<>();
        marked = new boolean[G.V()];
        for (int i = 0; i < G.V(); i++)
            marked[i] = false;

        visit(G, 0);
        while (!pq.isEmpty()) {
            Ex12Edge e = pq.delMin();
            int v = e.either(), w = e.other(v);
            if (marked[v] && marked[w]) continue;
            edges.add(e);
            weight += e.weight();
            if (!marked[v]) visit(G, v);
            if (!marked[w]) visit(G, w);
        }
    }

    @Override
    public Iterable<Ex12Edge> edges() {
        return edges;
    }

    @Override
    public double weight() {
        return weight;
    }

    private void visit(Ex13EdgeWeightedGraph G, int vertex) {
        marked[vertex] = true;
        for (Ex12Edge e : G.adj(vertex)) {
            int v = e.either(), w = e.other(v);
            if (!marked[v] || !marked[w])
                pq.insert(e);
        }
    }

    public static void main(String[] args) {
        In in = new In("tinyEWG.txt");
        Ex13EdgeWeightedGraph G = new Ex13EdgeWeightedGraph(in);
        StdOut.println(G);
        Ex16PrimsLazyMST mst = new Ex16PrimsLazyMST(G);
        StdOut.println("weight:" + mst.weight + " mst: ");
        for (Ex12Edge e : mst.edges())
            StdOut.println(e);
    }
}

File: PythonLearnig/Algorithms_Part_2/Ex16PrimsLazyMST.py
from Ex13EdgeWeightedGraph import EdgeWeightedGraph
from Ex15KruskalsMST import MinPQ


class PrimsLazyMST:

    def __init__(self, graph):
        self.all_edges = []
        self.total_weight = 0.0

        marked = [False] * graph.get_v()
        queue = MinPQ()

        def visit(vertex):
            marked[vertex] = True
            for edge in graph.get_adj(vertex):
                v = edge.either()
                w = edge.other(v)
                if marked[v] and marked[w]:
                    continue
                queue.put(edge)

        def execute():
            while not queue.is_empty():
                edge = queue.delete_min()
                v = edge.either()
                w = edge.other(v)
                if marked[v] and marked[w]:
                    continue
                self.all_edges.append(edge)
                self.total_weight += edge.get_weight()
                if marked[v]:
                    visit(w)
                else:
                    visit(v)

        visit(0)
        execute()

    def get_weight(self):
        return self.total_weight

    def edges(self):
        for edge in self.all_edges:
            yield edge


def test_prims_lazy_mst():

    graph = EdgeWeightedGraph(0)
    graph.init_from('tinyEWG.txt')
    print()
    print(graph)

    mst = PrimsLazyMST(graph)
    print(f'weight: {mst.get_weight()} mst:')
    for edge in mst.edges():
        print(f'\t{edge}')


if __name__ == '__main__':
    test_prims_lazy_mst()
